### ISSUE TITLE
Pluggable optimizer API

### ISSUE THREAD

**Issue by markshannon** [python/cpython#104584]
We need an API for optimizers to be plugged in to CPython.

The proposed model is that of client server, where the VM is the client and the optimizer is the server.
The optimizer registers with the VM, then VM calls the optimizer when hotspots are detected.

The API:

```C
type struct {
    OBJECT_HEADER;
    _PyInterpreterFrame *(*execute)(PyExecutorObject *self, _PyInterpreterFrame *frame, PyObject **stack_pointer);
    /* Data needed by the executor goes here, but is opaque to the VM */
} PyExecutorObject;

/* This would be nicer as an enum, but C doesn't define the size of enums */
#define PY_OPTIMIZE_FUNCTION_ENTRY 1
#define PY_OPTIMIZE_RESUME_AFTER_YIELD 2
#define PY_OPTIMIZE_BACK_EDGE 4
typedef uint32_t PyOptimizerCapabilities;

type struct {
    OBJECT_HEADER;
    PyExecutorObject *(*compile)(PyOptimizerObject* self, PyCodeObject *code, int offset);
    PyOptimizerCapabilities capabilities;
    float optimization_cost;
    float run_cost;
    /* Data needed by the compiler goes here, but is opaque to the VM */
} PyOptimizerObject;

void _Py_Executor_Replace(PyCodeObject *code, int offset, PyExecutorObject *executor);

int _Py_Optimizer_Register(PyOptimizerObject* optimizer);
```

The semantics of a `PyExecutorObject` is that upon return from its `execute` function, the VM state will have advanced `N` instructions. Where `N` is a non-negative integer.

Full discussion here: https://github.com/faster-cpython/ideas/discussions/380

This is not a replacement for PEP 523. That will need a PEP. We should get this working first, before we consider replacing PEP 523.

### Linked PRs
* gh-105100
* gh-105244
* gh-105683
* gh-105924
* gh-106126
* gh-106131
* gh-106141
* gh-106146
* gh-106163
* gh-106171
* gh-106277
* gh-106393
* gh-106484
* gh-106489
* gh-106492
* gh-106497
* gh-106500
* gh-106526
* gh-106641
* gh-106908
* gh-107513
* gh-108953
* gh-108961
* gh-109347
* gh-110593

**Comment by Bunos07** [comment]
Plsss

**Comment by Bunos07** [comment]
Plsss

**Comment by markshannon** [comment]
Note that the above API is just the initial version to support our work on speeding up Python 3.13.
It will probably need to extended to support PyTorch Dynamo and other users of PEP 523 that cannot use PEP 669, but that is for another issue.

**Comment by encukou** [comment]
Is documenting these planned for 3.13?

**Comment by markshannon** [comment]
Since we are removing them in 3.14, probably not.

**Comment by vstinner** [comment]
> Since we are removing them in 3.14, probably not.

If possible, I would like to backport https://github.com/python/cpython/pull/120643 to Python 3.13, to solve a C99 compatibility issue.

**Comment by erlend-aasland** [comment]
A bunch of PRs were merged; a quick glance at the PR list suggests that this issue can be closed as completed. Are there further work to be done?

**Comment by brandtbucher** [comment]
We discussed the optimizer API offline. In short: we should rip the API itself out, and just keep all of the code that does that actual optimizing.

The API itself is constantly changing, poorly-defined, and undocumented. There's a bunch of infrastructure required just to test the API (not the actual optimizations we perform), and it introduces indirection and artificial boundaries into some pretty performance-sensitive stuff. Nobody it using it that we're aware of, nobody we've talked to is planning on using it, and frankly we don't want anyone to *start* using it. So let's remove it.
